File: benchmarks/c/10_function_calls.c
// Function Call Overhead Benchmark - C Implementation

#include <stdio.h>

static inline long mix(long x, long y) {
    return ((x * 3) + (y * 5) - (x % 7) + (y % 11)) % 1000003;
}

int main() {
    const long n = 1500000;
    long acc = 0;

    for (long i = 0; i < n; i++) {
        acc = acc + mix(i, acc + i);
    }

    printf("function call checksum: %ld\n", acc);
    return 0;
}

File: benchmarks/rust/10_function_calls.rs
// Function Call Overhead Benchmark - Rust Implementation

#[inline(always)]
fn mix(x: i64, y: i64) -> i64 {
    ((x * 3) + (y * 5) - (x % 7) + (y % 11)) % 1_000_003
}

fn main() {
    let n = 1_500_000i64;
    let mut acc = 0i64;

    for i in 0..n {
        acc += mix(i, acc + i);
    }

    println!("function call checksum: {}", acc);
}
